src/shogun/kernel/PeriodicKernel.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 2015 Esben Soerig
+ */
+
+#include <shogun/lib/common.h>
+#include <shogun/base/Parameter.h>
+#include <shogun/features/DotFeatures.h>
+#include <shogun/io/SGIO.h>
+#include "PeriodicKernel.h"
+
+using namespace shogun;
+
+CPeriodicKernel::CPeriodicKernel() : CDotKernel()
+{
+	init();
+}
+
+CPeriodicKernel::CPeriodicKernel(int32_t size, float64_t w, float64_t p) : CDotKernel(size)
+{
+	init();
+	set_width(w);
+	set_period(p);
+}
+
+CPeriodicKernel::CPeriodicKernel(CDotFeatures* l, CDotFeatures* r,
+		float64_t w, float64_t p, int32_t size) : CDotKernel(size)
+{
+	init();
+	set_width(w);
+	set_period(p);
+	init(l,r);
+}
+
+CPeriodicKernel::~CPeriodicKernel()
+{
+	cleanup();
+}
+
+CPeriodicKernel* CPeriodicKernel::obtain_from_generic(CKernel* kernel)
+{
+	// TODO: Add proper periodic kernel enum value
+	if (kernel->get_kernel_type()!= K_PERIODIC)
+	{
+		SG_SERROR("CPeriodicKernel::obtain_from_generic(): provided kernel is "
+				"not of type CPeriodicKernel!\n");
+	}
+
+	/* since an additional reference is returned */
+	SG_REF(kernel);
+	return (CPeriodicKernel*)kernel;
+}
+
+void CPeriodicKernel::cleanup()
+{
+	if (sq_lhs != sq_rhs)
+		SG_FREE(sq_rhs);
+	sq_rhs = NULL;
+
+	SG_FREE(sq_lhs);
+	sq_lhs = NULL;
+
+	CKernel::cleanup();
+}
+
+void CPeriodicKernel::precompute_squared_helper(float64_t* &buf, CDotFeatures* df)
+{
+	ASSERT(df)
+	int32_t num_vec=df->get_num_vectors();
+	buf=SG_MALLOC(float64_t, num_vec);
+
+	for (int32_t i=0; i<num_vec; i++)
+		buf[i]=df->dot(i,df, i);
+}
+
+bool CPeriodicKernel::init(CFeatures* l, CFeatures* r)
+{
+	///free sq_{r,l}hs first
+	cleanup();
+
+	CDotKernel::init(l, r);
+	precompute_squared();
+	return init_normalizer();
+}
+
+float64_t CPeriodicKernel::compute(int32_t idx_a, int32_t idx_b)
+{
+	/* Periodic kernel defined as by David Duvenaud in 
+	 http://mlg.eng.cam.ac.uk/duvenaud/cookbook/index.html 
+
+	 k({\bf x},{\bf x'})= exp(-\frac{2sin^2(|{\bf x}-{\bf x'}|/p)}{l^2})
+	*/
+	return exp(-2*pow(sin(M_PI*distance(idx_a,idx_b)/period), 2)/pow(width,2));
+}
+
+void CPeriodicKernel::precompute_squared()
+{
+	if (!lhs || !rhs)
+		return;
+
+	precompute_squared_helper(sq_lhs, (CDotFeatures*) lhs);
+
+	if (lhs==rhs)
+		sq_rhs=sq_lhs;
+	else
+		precompute_squared_helper(sq_rhs, (CDotFeatures*) rhs);
+}
+
+SGMatrix<float64_t> CPeriodicKernel::get_parameter_gradient(
+		const TParameter* param, index_t index)
+{
+	REQUIRE(lhs && rhs, "Features not set!\n")
```

Comment:
Could you require those seperately? Users then have the information which side is NULL